stellargraph/mapper/mini_batch_node_generators.py

```diff
@@ -0,0 +1,361 @@
+        :meth:`keras.Model.predict_generator`,
+
+    This class should be created using the `.flow(...)` method of
+    :class:`ClusterNodeGenerator`.
+
+    Args:
+        graph (StellarGraph): The graph
+        clusters (list): A list of lists such that each sub-list indicates the nodes in a cluster.
+            The length of this list, len(clusters) indicates the number of batches in one epoch.
+        targets (np.ndarray, optional): An optional array of node targets of size (N x C),
+            where C is the target size (e.g., number of classes for one-hot class targets)
+        node_ids (iterable, optional): The node IDs for the target nodes. Required if targets is not None.
+        normalize_adj (bool, optional): Specifies whether the adjacency matrix for each mini-batch should
+            be normalized or not. The default is True.
+        q (int, optional): The number of subgraphs to combine for each batch. The default value is
+            1 such that the generator treats each subgraph as a batch.
+        lam (float, optional): The mixture coefficient for adjacency matrix normalisation (the
+            'diagonal enhancement' method). Valid values are in the interval [0, 1] and the default value is 0.1.
+        name (str, optional): An optional name for this generator object.
+    """
+
+    def __init__(
+        self,
+        graph,
+        clusters,
+        targets=None,
+        node_ids=None,
+        normalize_adj=True,
+        q=1,
+        lam=0.1,
+        name=None,
+    ):
+
+        self.name = name
+        self.clusters = list()
+        self.clusters_original = copy.deepcopy(clusters)
+        self.graph = graph
+        self.node_list = list(graph.nodes())
+        self.normalize_adj = normalize_adj
+        self.q = q
+        self.lam = lam
+        self.node_order = list()
+        self._node_order_in_progress = list()
+        self.__node_buffer = dict()
+
+        if len(clusters) % self.q != 0:
+            raise ValueError(
+                "The number of clusters should be exactly divisible by q. However {} is not exactly divisible by {}".format(
+                    len(clusters), q
+                )
+            )
+
+        if node_ids is not None:
+            self.target_ids = list(node_ids)
+
+        if targets is not None:
+            if node_ids is None:
+                raise ValueError(
+                    "Since targets is not None, node_ids must be given and cannot be None."
+                )
+
+            if (targets is not None) and (len(node_ids) != len(targets)):
+                raise ValueError(
+                    "When passed together targets and indices should be the same length."
+                )
+
+            self.target_ids = list(node_ids)
+            self.targets = np.asanyarray(targets)
+            self.target_node_lookup = dict(
+                zip(self.target_ids, range(len(self.target_ids)))
+            )
+            self.target_ids = list(node_ids)
```

Comment:
I count three times this assignment is (potentially) made. Some code rationalisation is in order.

Reply:
I've corrected this. Certainly not necessary to assign it 3 times!